### PR TITLE
feat: configurable fees receiver

### PR DESCRIPTION
**Pull request by Dodecahedr0x**
> ⚠️ NOTE: This PR changes the account structure for the Program Config
>
>  The `set_fees_receiver` migrates the Program Config and must be run immediately after the update to prevent blocking other instructions. Although the migration itself is tested, the behavior of all affected instructions is not tested with the old structure.

| Status  | Type  | ⚠️ Core Change | Issue |
| :---: | :---: | :---: | :--: |
| Ready| Feature | Yes | #89  |

## Problem

Enable permissionless protocol fee claiming.

## Solution

Stores the fee receiver in the ProgramConfig. Remove admin checks for claiming. Send protocol fees to the stored fee receiver.

## Deploy Notes

The `set_fees_receiver` instruction must be called by the admin when deploying this PR as other methods relying on the ProgramConfig would fail otherwise.

## Summary by CodeRabbit

* **New Features**
  * Add "Set Fees Receiver" instruction and configurable protocol fees receiver; claim flow now pays the configured receiver.
  * Introduce FeesVault account type with migration support and updated claim/rent semantics.
  * Add a fast instruction dispatch path for improved handling.

* **Tests**
  * New and updated integration/unit tests covering set-fees-receiver, claim flows, migration, and edge cases.

* **Chores**
  * Update test SDK and dependency versions.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

Updated On: 2025-09-15 16:53:17 UTC

This PR implements configurable fees receiver functionality that enables permissionless protocol fee claiming. The core change modifies the `ProgramConfig` struct to include a new `fees_receiver` field, allowing protocol fees to be sent to a designated address instead of always going to the program update authority.

Key architectural changes include:

**Account Structure Changes:**
- Extended `ProgramConfig` with a `fees_receiver: Pubkey` field (+32 bytes)
- Updated size calculations and serialization logic across the codebase
- Implemented migration logic in `process_set_fees_receiver` to handle existing accounts

**Protocol Fee Claiming Refactor:**
- Removed admin authorization requirements from `protocol_claim_fees`
- Changed function signature from `protocol_claim_fees(admin: Pubkey)` to `protocol_claim_fees(fees_receiver: Pubkey, program: Pubkey)`
- Updated account structure to include program config and fees receiver accounts
- Fees now transfer to the configured receiver instead of the program authority

**New Instruction Implementation:**
- Added `SetFeesReceiver` discriminator (value 16) and corresponding processor
- Created `SetFeesReceiverArgs` struct to pass the new fees receiver address
- Implemented admin-only `set_fees_receiver` instruction that migrates existing configs

**Integration Changes:**
- Updated test fixtures and integration tests to support the new account structure
- Switched from local file dependencies to published npm packages (`@magicblock-labs/ephemeral-rollups-sdk` v0.2.10)
- Modified test delegation programs to explicitly specify validators

The change enables automatic fee distribution systems and makes the protocol more adaptable for staking program integration. The migration mechanism ensures backward compatibility during deployment by detecting old account formats and upgrading them with default values.

## Confidence score: 3/5

- This PR introduces significant breaking changes that require careful deployment and immediate migration to prevent blocking other instructions
- Score reflects the complexity of account structure migration and the critical nature of fee handling logic changes
- Pay close attention to migration logic in `src/processor/set_fees_receiver.rs` and the updated fee claiming mechanism in `src/processor/protocol_claim_fees.rs`

### CODE REVIEW

File: tests/test_set_fees_receiver.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    let ix = dlp::instruction_builder::set_fees_receiver(admin.pubkey(), fees_receiver);
+    let tx = Transaction::new_signed_with_payer(
+        &[ix],
+        Some(&payer.pubkey()),
+        &[&payer, &admin],
+        blockhash,
+    );
+    let res = banks.process_transaction(tx).await;
+    assert!(res.is_ok());
+
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply tests/test_set_fees_receiver.rs]: The goal here is explicitly to test migration

File: tests/test_protocol_claim_fees.rs [hunk 2]
```diff
@@ -19,34 +21,27 @@ async fn test_protocol_claim_fees() {
...
-    let ix = dlp::instruction_builder::protocol_claim_fees(admin.pubkey());
-    let tx = Transaction::new_signed_with_payer(
-        &[ix],
-        Some(&payer.pubkey()),
-        &[&payer, &admin],
-        blockhash,
-    );
+    let ix = dlp::instruction_builder::protocol_claim_fees(admin.pubkey(), dlp::ID);
+    let tx = Transaction::new_signed_with_payer(&[ix], Some(&payer.pubkey()), &[&payer], blockhash);
     let res = banks.process_transaction(tx).await;
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply tests/test_protocol_claim_fees.rs]: Addressed in [90e02a13aa0887772c2a4ed17c6e5a483fdf77d9](https://github.com/magicblock-labs/delegation-program/pull/90/commits/90e02a13aa0887772c2a4ed17c6e5a483fdf77d9)

File: src/cu.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        log!("BENCHMARK BEGIN: [{}]", name);
+        Self {
+            name,
+            remaining_at_start: Self::remaining_cu(),
+        }
+    }
+
+    fn remaining_cu() -> u64 {
+        unsafe { sol_remaining_compute_units() }
+    }
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/cu.rs]: Comes from merging main commits, not relevant to this PR

File: src/processor/fast/commit_state_from_buffer.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+) -> ProgramResult {
+    let [validator, delegated_account, commit_state_account, commit_record_account, delegation_record_account, delegation_metadata_account, state_buffer_account, validator_fees_vault, program_config_account, _system_program] =
+        accounts
+    else {
+        return Err(ProgramError::NotEnoughAccountKeys);
+    };
+
+    let args =
+        CommitStateFromBufferArgs::try_from_slice(data).map_err(|_| ProgramError::BorshIoError)?;
+
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/fast/commit_state_from_buffer.rs]: Comes from merging main commits, not relevant to this PR

File: src/processor/fast/commit_state.rs [hunk 5]
```diff
@@ -52,14 +57,14 @@ pub fn process_commit_state(
     accounts: &[AccountInfo],
     data: &[u8],
 ) -> ProgramResult {
-    let args = CommitStateArgs::try_from_slice(data)?;
+    let args = CommitStateArgs::try_from_slice(data).map_err(|_| ProgramError::BorshIoError)?;
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/fast/commit_state.rs]: Comes from merging main commits, not relevant to this PR

File: src/processor/fast/delegate.rs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    else {
+        return Err(ProgramError::NotEnoughAccountKeys);
+    };
+
+    require_owned_pda(delegated_account, &crate::fast::ID, "delegated account")?;
+
+    // Check that payer and delegated_account are signers, this ensures the instruction is being called from CPI
+    require_signer(payer, "payer")?;
+    require_signer(delegated_account, "delegated account")?;
+
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/fast/delegate.rs]: Comes from merging main commits, not relevant to this PR

File: src/processor/fast/finalize.rs [hunk 7]
```diff
@@ -50,54 +53,59 @@ pub fn process_finalize(
...
 
     // Check that the commit record is the right one
-    if !commit_record.account.eq(delegated_account.key) {
+    if !pubkey_eq(commit_record.account.as_array(), delegated_account.key()) {
         return Err(DlpError::InvalidDelegatedAccount.into());
     }
-    if !commit_record.identity.eq(validator.key) {
+    if !pubkey_eq(commit_record.identity.as_array(), validator.key()) {
         return Err(DlpError::InvalidReimbursementAccount.into());
     }
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/fast/finalize.rs]: Comes from merging main commits, not relevant to this PR

File: src/processor/fast/mod.rs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+pub use commit_state_from_buffer::*;
+pub use delegate::*;
+pub use finalize::*;
+pub use undelegate::*;
+
+pub fn to_pinocchio_program_error(
+    error: solana_program::program_error::ProgramError,
+) -> pinocchio::program_error::ProgramError {
+    u64::from(error).into()
+}
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/fast/mod.rs]: Comes from merging main commits, not relevant to this PR

File: src/processor/fast/utils/requires.rs [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+    }
+    require_pda(
+        program_config,
+        &[pda::PROGRAM_CONFIG_TAG, program],
+        &crate::fast::ID,
+        is_writable,
+        "program config",
+    )?;
+    Ok(!pubkey_eq(program_config.owner(), &pinocchio_system::ID))
+}
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/fast/utils/requires.rs]: Comes from merging main commits, not relevant to this PR

File: tests/integration/tests/test-delegation.ts [hunk 10]
```diff
@@ -223,6 +317,25 @@ describe("TestDelegation", () => {
...
 
+  it("Set fees receiver", async () => {
+    const ix = createSetFeesReceiverInstruction(
+      admin,
+      admin,
+      testDelegation.programId
+    );
+    const txId = await processInstruction(ix);
+    console.log("Set fees receiver tx:", txId);
+  });
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply tests/integration/tests/test-delegation.ts]: Addressed in [584d0bc15ab474e03333630dfa46ab084ff99b5f](https://github.com/magicblock-labs/delegation-program/pull/90/commits/584d0bc15ab474e03333630dfa46ab084ff99b5f)

File: src/processor/set_fees_receiver.rs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    };
+
+    if migrated {
+        resize_pda(
+            admin,
+            fees_vault_account,
+            system_program,
+            fees_vault.size_with_discriminator(),
+        )?;
+    }
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/set_fees_receiver.rs]: Addressed in [364abb2c8d41c681aa41f57b803f7aa253b69ba1](https://github.com/magicblock-labs/delegation-program/pull/90/commits/364abb2c8d41c681aa41f57b803f7aa253b69ba1)

File: src/processor/utils/curve.rs [hunk 12]
```diff
@@ -1,6 +1,26 @@
...
+        let ret = unsafe {
+            pinocchio::syscalls::sol_curve_validate_point(
+                0, // 0 means Ed25519
+                key.as_ptr(),
+                &mut result as *mut u8,
+            )
+        };
+
+        ret == 0
+    }
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/utils/curve.rs]: Comes from merging main commits, not relevant to this PR

File: src/processor/protocol_claim_fees.rs [hunk 13]
```diff
@@ -28,40 +28,42 @@ pub fn process_protocol_claim_fees(
     _data: &[u8],
 ) -> ProgramResult {
     // Load Accounts
-    let [admin, fees_vault, delegation_program_data] = accounts else {
+    let [fees_vault_account, fees_receiver, _program] = accounts else {
         return Err(ProgramError::NotEnoughAccountKeys);
     };
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/protocol_claim_fees.rs]: Addressed in [e42869fd656c112d9128cb4594ecfc3b703dac57](https://github.com/magicblock-labs/delegation-program/pull/90/commits/e42869fd656c112d9128cb4594ecfc3b703dac57)

File: tests/integration/tests/test-delegation.ts [hunk 14]
```diff
@@ -321,6 +317,18 @@ describe("TestDelegation", () => {
...
+    const ix = createSetFeesReceiverInstruction(admin, admin);
+    const txId = await processInstruction(ix);
+    console.log("Set fees receiver tx:", txId);
+  });
+
+  it("Claim protocol fees", async () => {
+    const ix = createClaimProtocolFeesVaultInstruction(admin);
+    const txId = await processInstruction(ix);
+    console.log("Claim protocol fees tx:", txId);
+  });
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply tests/integration/tests/test-delegation.ts]: Out of scope for this PR as these verifications are never done in the test

File: src/processor/protocol_claim_fees.rs [hunk 15]
```diff
@@ -1,67 +1,68 @@
...
 
     // Calculate the amount to transfer
-    let min_rent = Rent::default().minimum_balance(8);
-    if fees_vault.lamports() < min_rent {
+    let min_rent = Rent::get()?.minimum_balance(fees_vault.size_with_discriminator());
+    if fees_vault_account.lamports() < min_rent {
         return Err(ProgramError::InsufficientFunds);
     }
-    let amount = fees_vault.lamports() - min_rent;
+    let amount = fees_vault_account.lamports() - min_rent;
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/protocol_claim_fees.rs]: Addressed in [7b1fb92105dce0308e5cd73f3fbf38fb0f469157](https://github.com/magicblock-labs/delegation-program/pull/90/commits/7b1fb92105dce0308e5cd73f3fbf38fb0f469157)

File: src/processor/protocol_claim_fees.rs [hunk 16]
```diff
@@ -1,67 +1,71 @@
...
+    **fees_vault_account.try_borrow_mut_lamports()? = fees_vault_account
         .lamports()
         .checked_sub(amount)
         .ok_or(ProgramError::InsufficientFunds)?;
 
-    **admin.try_borrow_mut_lamports()? = admin
+    **fees_receiver.try_borrow_mut_lamports()? = fees_receiver
         .lamports()
         .checked_add(amount)
         .ok_or(ProgramError::ArithmeticOverflow)?;
```
[the remark this answers is not in the document]
Comment (Dodecahedr0x) [reply src/processor/protocol_claim_fees.rs]: Addressed in [d974afbddeb89eb9394478328d11ff42139a32f6](https://github.com/magicblock-labs/delegation-program/pull/90/commits/d974afbddeb89eb9394478328d11ff42139a32f6)
